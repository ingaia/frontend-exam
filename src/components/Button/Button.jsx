import React from 'react';
import PropTypes from 'prop-types';
import { ButtonStyled } from './styled'

const Button = ({type, className, disabled, name, text, onClick}) => {
  return (
    <ButtonStyled type={type} disabled={!!disabled} name={name} className={className} onClick={onClick}>
      {text}
    </ButtonStyled>
  )
};

export default Button

Button.propTypes = {
  className: PropTypes.any,
  disabled: PropTypes.any,
  name: PropTypes.any.isRequired,
  type: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired
}
