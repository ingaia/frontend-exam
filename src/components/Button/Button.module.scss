@import 'scss/components.scss';

/* */

.button {

    $min-size: 40px;
    $min-size-small: 30px;

    /* */

    position: relative;

    min-width: $min-size;
    min-height: $min-size;

    background: transparent;

    font-family: 'Roboto Condensed', sans-serif;
    @include font-sizer(14);
    letter-spacing: 1px;
    text-transform: uppercase;
    color: white;

    cursor: pointer;

    outline: none !important;

    border: none;

    z-index: 1;

    backface-visibility: hidden;

    :global(.theme--dark--high-contrast) & {

        color: black !important;

    }

    &--full {

        width: 100%;

    }

    &--padding {

        padding: 0 $theme-padding * 2;

    }

    &--small {

        min-height: $min-size-small

    }

    &--loading {

        color: transparent !important;

        border-color: inherit;

        &::before {

            content: '';

            $size: 20px;

            position: absolute;

            top: 50%;
            left: 50%;

            width: $size;
            height: $size;

            margin-top: -$size / 2;
            margin-left: -$size / 2;

            border: 3px solid transparent;
            border-right-color: inherit;

            border-radius: $size;

            animation: rotate .4s linear infinite;

            @keyframes rotate {

                from {

                    transform: rotate(0deg);

                }

                to {

                    transform: rotate(360deg);

                }

            }

        }

    }

    &__background {

        position: absolute;

        top: 0;
        left: 0;

        width: 100%;
        height: 100%;

        z-index: -1;

        border: 1px solid transparent;

    }

    &:hover &__background {

        filter: brightness(0.9);

    }

    &:active &__background {

        filter: brightness(0.8);

    }

    &:not(:last-of-type){

        margin-right: $theme-padding;

    }

}
