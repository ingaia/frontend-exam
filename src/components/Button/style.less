@import './../../variables.less';

.btn {
    padding: 10px 30px;
    margin-bottom: 20px;
    background-color: transparent;
    font-size: 14px;
    cursor: pointer;

    &.btn-hillary {
        border: 1px solid @hillary;
        color: @hillary;
    }

    &.btn-default {
        border: none;
        color: #fff;
        font-weight: 100;   
    }
}