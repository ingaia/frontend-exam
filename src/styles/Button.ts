import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(() => ({
  root: (props: any) => ({
    color: 'var(--hillary)',
    backgroundColor: 'transparent',
    border: props.noBorders ? 'unset' : '1px solid var(--hillary)',
    borderRadius: 0,
    marginTop: '30px'
  })
}));
