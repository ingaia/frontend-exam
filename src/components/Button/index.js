import React from 'react';

import './style.less';

export default function Button(props) {
    return (
        <button 
            className={`btn ${props.className}`}
            type={props.type ? props.type : 'button'} 
            onClick={props.onClick}>
                {props.label ? props.label.toUpperCase() : ''}
        </button>
    );
}