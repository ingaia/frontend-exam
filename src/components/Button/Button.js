import React from 'react'
import PropTypes from 'prop-types'

/* */

import styles from './Button.module.scss'

/* */

class Button extends React.Component {

    colorDefault(){

        return /\d/.test(this.props.color) || this.props.color === 'white' || this.props.color === 'black' ? this.props.color : this.props.color + '-500'

    }

    withLabel(){

        return (this.props.label !== '' && true) || ''

    }

    /* */

    render(){

        const colorDefault = this.colorDefault()
        const withLabel = this.withLabel()

        /* */

        return (

            <button

            onClick={ this.props.onClick }

            className={

                [

                    styles.button,
                    withLabel && styles['button--padding'],
                    this.props.full && styles['button--full'],
                    this.props.small && styles['button--small'],
                    this.props.loading && styles['button--loading']

                ].join(' ')

            } type={ this.props.type } style={{

                color: `var(--md-${colorDefault})`,
                borderColor: this.props.light ? `var(--md-grey-200)` : `var(--md-${colorDefault})`

            }}>

                <div className={ styles['button__background'] } style={{

                    borderColor: this.props.light ? `var(--md-grey-200)` : `var(--md-${colorDefault})`

                }}></div>

                { this.props.label }

            </button>

        )

    }

}

Button.propTypes = {

    disabled: PropTypes.bool,
    loading : PropTypes.bool,

    color   : PropTypes.string,

    light   : PropTypes.bool,
    icon    : PropTypes.string,
    full    : PropTypes.bool,
    small   : PropTypes.bool,

    label   : PropTypes.string,
    type    : PropTypes.string.isRequired

}

Button.defaultProps = {

    type: 'button',

    color: 'blue-500'

}

/* */

export default Button
