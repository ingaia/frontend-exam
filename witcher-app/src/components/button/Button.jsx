import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import styles from './Button.module.scss';

const Button = (props) => {

    const { label, className, onClick } = props;

    return (
        <button 
            { ...props }
            className={classnames(styles.Button, className)}
            onClick={onClick}
        >
            {label}
        </button>
    );
}
 
Button.propTypes = {
    label: PropTypes.string,
    className: PropTypes.string,
    onClick: PropTypes.func,
}

Button.defaultProps = {
    className: 'Button',
    onClick: () => {},
}

export default Button;