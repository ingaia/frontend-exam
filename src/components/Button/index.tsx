import React from 'react';
import Button from '@material-ui/core/Button';

import { useStyles } from '../../styles/Button';

export interface Props {
  children: any,
  type?: "button" | "reset" | "submit",
  fullWidth?: boolean,
  noborders?: boolean,
  onClick?: any
}

const Component = ({children, noborders, ...props }: Props) => (
  <Button
    {...props}
    classes={useStyles({ noBorders: noborders })}>
    {children}
  </Button>
);

export default Component;
