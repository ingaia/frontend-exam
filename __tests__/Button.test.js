import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json'
import Button from '../src/components/Button';

describe('<Button/>', () => {
    it('should render', () => {
        const wrapper = shallow(<Button label="Label"/>);
        
        expect(wrapper.exists('[type="button"]')).toBe(true);
        expect(wrapper.hasClass('btn')).toBe(true);
        expect(wrapper.text()).toBe("LABEL");
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    }); 

    it('should render the button with type "submit"', () => {
        const wrapper = shallow(<Button type="submit"/>);
        
        expect(wrapper.exists('[type="submit"]')).toBe(true);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    }); 
});