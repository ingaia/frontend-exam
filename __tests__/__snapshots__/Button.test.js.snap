// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render 1`] = `
<button
  className="btn undefined"
  type="button"
>
  LABEL
</button>
`;

exports[`<Button/> should render the button with type "submit" 1`] = `
<button
  className="btn undefined"
  type="submit"
/>
`;
